packages/fx-core/src/client/teamsDevPortalClient.ts

```diff
@@ -0,0 +1,947 @@
+  ErrorMessages,
+} from "../component/driver/teamsApp/constants";
+import { AppStudioError } from "../component/driver/teamsApp/errors";
+import {
+  ApiSecretRegistration,
+  ApiSecretRegistrationUpdate,
+} from "../component/driver/teamsApp/interfaces/ApiSecretRegistration";
+import { AsyncAppValidationDetailsResponse } from "../component/driver/teamsApp/interfaces/AsyncAppValidationDetailsResponse";
+import { AsyncAppValidationResponse } from "../component/driver/teamsApp/interfaces/AsyncAppValidationResponse";
+import { AsyncAppValidationResultsResponse } from "../component/driver/teamsApp/interfaces/AsyncAppValidationResultsResponse";
+import { OauthConfigurationId } from "../component/driver/teamsApp/interfaces/OauthConfigurationId";
+import { OauthRegistration } from "../component/driver/teamsApp/interfaces/OauthRegistration";
+import { IPublishingAppDenition } from "../component/driver/teamsApp/interfaces/appdefinitions/IPublishingAppDefinition";
+import { IValidationResult } from "../component/driver/teamsApp/interfaces/appdefinitions/IValidationResult";
+import { AppDefinition } from "../component/driver/teamsApp/interfaces/appdefinitions/appDefinition";
+import { AppUser } from "../component/driver/teamsApp/interfaces/appdefinitions/appUser";
+import { AppStudioResultFactory } from "../component/driver/teamsApp/results";
+import { manifestUtils } from "../component/driver/teamsApp/utils/ManifestUtils";
+import {
+  BotChannelType,
+  IBotRegistration,
+} from "../component/resource/botService/appStudio/interfaces/IBotRegistration";
+import { isHappyResponse } from "../component/resource/botService/common";
+import { TeamsFxUrlNames } from "../component/resource/botService/constants";
+import {
+  BotFrameworkConflictResultError,
+  BotFrameworkForbiddenResultError,
+  BotFrameworkNotAllowedToAcquireTokenError,
+  BotRegistrationNotFoundError,
+  ConfigUpdatingError,
+  ProvisionError,
+} from "../component/resource/botService/errors";
+import { Messages } from "../component/resource/botService/messages";
+import { CommonStrings, ConfigNames } from "../component/resource/botService/strings";
+import {
+  CheckSideloadingPermissionFailedError,
+  DeveloperPortalAPIFailedError,
+} from "../error/teamsApp";
+
+export class RetryHandler {
+  public static RETRIES = 6;
+  public static async Retry<T>(fn: () => Promise<T>): Promise<T | undefined> {
+    let retries = this.RETRIES;
+    let response;
+    while (retries > 0) {
+      retries = retries - 1;
+      try {
+        response = await fn();
+        return response;
+      } catch (e: any) {
+        // Directly throw 404 error, keep trying for other status code e.g. 503 400
+        if (retries <= 0 || e.response?.status == 404 || e.response?.status == 409) {
+          throw e;
+        } else {
+          await new Promise((resolve) => setTimeout(resolve, 5000));
+        }
+      }
+    }
+  }
+}
+
+class TeamsDevPortalClient {
+  endpoint: string;
+  region?: string;
```

Comment:
Can we call them `globalEndpoint `and `regionalEndpoint`? Normally, `region `should be a country code, not a regional url.